modules/fbs/includes/fbs.user.inc

```diff
@@ -11,22 +11,19 @@
  * @param string $name
  *   User name.
  * @param string $pincode
- *   Supplied pincode.
+ *   Supplied pin code.
+ * @param bool $sso
+ *   Used to indicate if this is single sign on and preAuthenticate should be used.
  *
  * @return array
  *   Authentication result.
+ *
+ * @throws DingProviderUserAuthFailure
+ *   If authentication fails.
  */
-function fbs_user_authenticate($name, $pincode) {
+function fbs_user_authenticate($name, $pincode, $sso = FALSE) {
   $res = new stdClass();
-  $pass = _fbs_wayf_login($name);
-  if ($pass) {
-    // Wayf login.
-    if ($pass !== $pincode) {
-      // Die if we didn't get the same pincode as we generated in
-      // wayf_login_credentials.
-      throw new DingProviderUserAuthFailure();
-    }
-
+  if ($sso) {
```

Comment:
Instead of keeping this conditional and expanding the signature of the callback could we instead implement this in the auth provider?

Reply:
I can't see how... it would require us to change this into two authentication functions one for single sign on (FBS PreAuthenticated) and one for normal login (authenticate).

Which then would call an internal function to do the work that the two shares. I think that would change more that the one parameter with default value.

Reply:
You just explained exactly how I would do it. 

I do not think it is a matter of changing more or less code - but a matter of keeping code consistent. The problem with the current approach is twofold:

1. You are adding a new parameter `$sso`to a existing well-defined provider function. I recognize that is has a default parameter but still.
2. You are making the `$pincode` parameter obsolete when using `$sso`. In general authentication is not needed when logging in using SSO. This has already been handled by another system.

I agree you would end up with three functions:

```php
fbs_user_authenticate($name, $pincode)
fbs_auth_authenticate($uuid)
_fbs_authenticate_patron(Patron $patron)
```

Each of these functions would have a more clear responsibility. We would even be able to unit test `_fbs_authenticate_patron` as it does not have any dependencies on external services.

I do not think that is worse.

If you decide to expand the signature for the user authenticate provider then you should at least document it by updating other providers as well with the new argument.

Reply:
As suggested... so created :-)

I have not updated the alma, openruth provider as there is a ticket on remove this code (the systems don't exists anymore).

Also added support in connie provider (and removed wayf support).